### PR TITLE
Add Option to Skip Flushing in TableBuilder

### PR DESCRIPTION
**Pull request by SherlockNoMad**
See https://github.com/facebook/rocksdb/issues/792

@sying, as we have discussed, I have added a new option skip_table_builder_flush
In addition, I make the buffer size configurable through option. This allows user to tune for optimal disk I/O size for flush.

### CODE REVIEW

File: include/rocksdb/env.h [hunk 1]
```diff
@@ -94,6 +94,12 @@ struct EnvOptions {
   // See DBOPtions doc
   size_t random_access_max_buffer_size;
 
+  // See DBOptions doc
+  size_t writable_file_max_buffer_size = 1024 * 1024;
+
+  // See DBOptions doc
+  bool skip_table_builder_flush = false;
```
Comment (siying) [inline include/rocksdb/env.h]: I prefer put this skip flush option in BlockBasedTableOptions

Comment (siying) [inline include/rocksdb/env.h]: This is doesn't not belong to EnvOptions. It's arguable should be CFOptions or BlockBasedTableOptions, but Env should not have any knowledge of "table".

File: table/block_based_table_builder.cc [hunk 2]
```diff
@@ -588,18 +593,22 @@ void BlockBasedTableBuilder::Add(const Slice& key, const Slice& value) {
...
+  }
+}
+
+void BlockBasedTableBuilder::SealDataBlock() {
+  Rep* r = rep_;
   assert(!r->closed);
   if (!ok()) return;
   if (r->data_block.empty()) return;
   WriteBlock(&r->data_block, &r->pending_handle);
-  if (ok()) {
```
Comment (siying) [inline table/block_based_table_builder.cc]: Can we keep the original logic here, and do "if (ok() && !r->skip_flush) {"?

File: include/rocksdb/env.h (hunk 1 above)
Comment (SherlockNoMad) [inline include/rocksdb/env.h]: OK. I will put it under BlockBasedTableOptions.

File: table/block_based_table_builder.cc (hunk 2 above)
Comment (SherlockNoMad) [inline table/block_based_table_builder.cc]: fixed.

**Comment by SherlockNoMad** [comment]
Hi @siying, really appreciate it if you can take a look at this...

**Comment by igorcanadi** [comment]
@SherlockNoMad Siying is currently on a vacation, he will be back next week.

**Comment by IslamAbdelRahman** [comment]
@igorcanadi, I will take a look

File: table/block_based_table_builder.cc [hunk 3]
```diff
@@ -592,14 +592,14 @@ void BlockBasedTableBuilder::Flush() {
...
+  if (ok() && !r->table_options.skip_table_builder_flush) {
     r->status = r->file->Flush();
   }
   if (r->filter_block != nullptr) {
     r->filter_block->StartBlock(r->offset);
   }
   r->props.data_size = r->offset;
-  ++r->props.num_data_blocks;
+  ++r->props.num_data_blocks;
```
Comment (IslamAbdelRahman) [inline table/block_based_table_builder.cc]: Can you please remove the space? this make the linter complain.
If you are using arcanist, you can run "arc lint"

**Comment by IslamAbdelRahman** [comment]
LGTM, will merge after fixing the minor comment

**Comment by SherlockNoMad** [comment]
Thank you @IslamAbdelRahman, I have fixed the white space at the end of line.

**Pull request merged**